pkg/binary/envoy/fetch_test.go

```diff
@@ -15,31 +15,163 @@
 
+func TestUntarEnvoyError(t *testing.T) {
+	tempDir, removeTempDir := morerequire.RequireNewTempDir(t)
+	dst := filepath.Join(tempDir, "dst")
+	defer removeTempDir()
+
+	var realHandler func(w http.ResponseWriter, r *http.Request)
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if realHandler != nil {
+			realHandler(w, r)
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+		}
+	}))
+	defer server.Close()
+
+	url := server.URL + "/file.tar.xz"
+	t.Run("error on incorrect URL", func(t *testing.T) {
+		e := untarEnvoy(dst, url, io.Discard)
+		require.EqualError(t, e, fmt.Sprintf(`received 404 status code from %s`, url))
+	})
+
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(200)
+	}
+	t.Run("error on empty", func(t *testing.T) {
+		e := untarEnvoy(dst, url, io.Discard)
+		require.EqualError(t, e, fmt.Sprintf(`not a valid xz stream %s: unexpected EOF`, url))
+	})
+
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(200)
+		w.Write([]byte("mary had a little lamb")) //nolint
+	}
+	t.Run("error on not a tar", func(t *testing.T) {
+		e := untarEnvoy(dst, url, io.Discard)
+		require.EqualError(t, e, fmt.Sprintf(`not a valid xz stream %s: xz: invalid header magic bytes`, url))
+	})
+
+	// Make the handler return the tar.xz
+	realHandler = func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(200)
+		zw, err := xz.NewWriter(w)
+		require.NoError(t, err)
+		defer zw.Close() //nolint
+		require.NoError(t, os.Mkdir(filepath.Join(tempDir, "envoy-v2"), 0750))
+		err = tar.Tar(zw, tempDir, "envoy-v2")
+		require.NoError(t, err)
+	}
+}
+
+func TestVerifyEnvoy(t *testing.T) {
+	tempDir, removeTempDir := morerequire.RequireNewTempDir(t)
+	defer removeTempDir()
+
+	platformPath := filepath.Join(tempDir, "platform")
+	require.NoError(t, os.MkdirAll(filepath.Join(platformPath, "bin"), 0755))
+	t.Run("envoy binary doesn't exist", func(t *testing.T) {
+		envoyPath, e := verifyEnvoy(platformPath)
+		require.Empty(t, envoyPath)
+		require.Contains(t, e.Error(), "no such file or directory")
+	})
+
+	expectedEnvoyPath := filepath.Join(platformPath, "bin", "envoy")
+	require.NoError(t, os.WriteFile(expectedEnvoyPath, []byte{}, 0600))
+	t.Run("envoy binary not executable", func(t *testing.T) {
+		envoyPath, e := verifyEnvoy(platformPath)
+		require.Empty(t, envoyPath)
+		require.EqualError(t, e, fmt.Sprintf(`envoy binary not executable at %q`, expectedEnvoyPath))
+	})
+
+	require.NoError(t, os.Chmod(expectedEnvoyPath, 0750))
+	t.Run("envoy binary ok", func(t *testing.T) {
+		envoyPath, e := verifyEnvoy(platformPath)
+		require.Equal(t, expectedEnvoyPath, envoyPath)
+		require.Nil(t, e)
+	})
+}
+
+func TestUntarEnvoy(t *testing.T) {
+	tests := []struct {
+		extension string
+		path      string
+	}{
+		{
+			extension: "tar.xz", // As of May 2021 all releases were compressed using xz
```

Comment:
now realistic examples, notably the basepath in the real life archives is related to the archive name